Design/DesignDoc-Least-ConnectionsTCPLB

```diff
@@ -0,0 +1,181 @@
+# Design document
+
+As requested, the project is broken into 2 components
+- the TCP load balancer library
+- a server that exposes the load balancer functionality and provides secure communication
+
+The load balancer is written in Go.
+
+Here is a simplified high level diagram:
+```mermaid
+graph TD;
+  user(User) -->|mTLS| auth_scheme(Auth Scheme)
+  auth_scheme --> rate_limiter(Rate Limiter)
+  rate_limiter --> least_conn(Least Connection)
+  least_conn --> upstream1(Upstream 1)
+  least_conn --> upstream2(Upstream 2)
+  subgraph server[Server]
+    auth_scheme
+    rate_limiter
+    least_conn
+    subgraph forwarder[Forwarder]
+      rate_limiter
+      least_conn
+    end
+  end
+  class user person;
+```
+
+## Requirements
+- certificates for the client are generated my the same instance as for the server. That permits to set the CN values adequately.
+- the CN format used is the host name + domain name
+
+## Load balancer library
+The load balancer library will contain the core logic for distributing incoming requests across backend servers. It will:
+- provide methods to forward a TCP connection to the upstream servers of the load balancer
+- use a least connection algorithm which means we will keep track of the usage of the upstream servers
+- rate limit the connections based on the client usage
+
+### Connection forwarder
+The forwarder maintains a list of connections opened towards the upstream servers. It will forward incoming TCP connections to the least used upstream server based on the least connection algorithm.
+
+The forwarder provides 2 public methods
+```go
+type Forwarder struct {
+	upstreams map[string]int
+	mu        sync.Mutex
+}
+func NewForwarder(upstreams []string) *Forwarder {}
+func (f *Forwarder) Forward(src net.Conn, allowedUpstreams []string) {}
+```
+
+The map of upstreams containing the connections count uses a mutex to make it thread safe.
+
+Note: the usage of mutex can be replaced by using channels.
+Some advantages of mutex are 
+- simpler implementation
+- light weight locking
+- low memory overhead
+Channels would permit passing parameters using them and can permit multiple consumers and producers concurrently which is not needed here.
+That choice can be made during implementation.
+
+After finding the least used upstream server, it increments the count and opens a new connection towards that server. It copies the data from the client connection to the upstream connection and vice versa using `io.Copy`.
+
+Copying the data in 2 different goroutines (one from client to server, one from server to client) ensures non blocking behavior.
+
+The forwarder manages the connections availability by using a retry mechanism. If connecting to the selected upstream fails, it will try the next least used server until it succeeds or exhausts the allowed upstreams list.
```

Comment:
Can you elaborate on this? Base on what will you decide that the connection failed?

Reply:
I have based this on the algorithm to be implemented.